File: GCJ/SavingTheUniverse/Nicholas/gcj01.cc
/* 
Google Code Jam Entrance Test 
(Saving the universe: https://code.google.com/codejam/contest/32013/dashboard)

Used Algorithm Method: Dynamic Programming
Coded by isdemus (May 01 2016)
*/

#include <iostream>
#include <map>
using namespace std;

int main(int argc, char* argv[]) {
	int testCases, searchEngine, queries;
	string line;
	cin >> testCases;
	getline(cin, line); // end of line


	for (int N=1; N<=testCases; N++) {
		cin >> searchEngine;
		getline(cin, line); // end of line

		map<string, int> engines;
		for (int s=0; s<searchEngine; s++) {
			getline(cin, line); // search engine
			engines[line] = s;
		}

		cin >> queries;
		//cout << "queries: " << queries << endl;
		int queriesArray[queries];
		int optimalSwitch[queries];

		getline(cin, line);
		for (int q=0; q<queries; q++) {
			getline(cin, line);
			queriesArray[q] = engines[line];
		}

		// Consider a system with various search engines as a blackbox
		// It does not matter which statement we use, it matters with the optimum (minimum) number of switches required
		for (int q=0; q<queries; q++) { // For each query, we find/use the optimal switch value from a previous query
			optimalSwitch[q] = queries-1; // begin with the max number of switches required <-> Guaranteed way to save the universe
			for (int s=0; s<searchEngine; s++) { // We apply a query to each search engine to see the compatibility
				if (queriesArray[q] == s) continue; // When we have a query that is the same name of the search engine
				else {
					int qPointer = q;
					for (qPointer; qPointer >= 0 && queriesArray[qPointer] != s; qPointer--) { /* intentionally spaced blank */ }
					if (qPointer == -1)
						optimalSwitch[q] = 0;
					else
						optimalSwitch[q] = min(optimalSwitch[q], optimalSwitch[qPointer] + 1); // Recurrence formula
				}
			}
		}
		

		cout << "Case #" << N << ": " << optimalSwitch[queries-1] << endl;
	}

	return 0;
}

File: GCJ/SavingTheUniverse/Nicholas/gcj01.py
#!/usr/bin/python

# Google Code Jam Entrance Test 
# (Saving the universe: https://code.google.com/codejam/contest/32013/dashboard)
# Used greedy algorithm (probably not optimal)
# Coded by isdemus (2016 05 03)

import sys

print("This is a GCJ01 problem" +
      "and this is coded in Python 3.5 environment")

def find_min_search_engine(S_list, Q_list, begin, current):
    answer_engine = current
    minimum = len(Q_list) - begin
    for search_engine in S_list:
        candidate = Q_list[begin:].count(search_engine)
        print("min: ", minimum,  " and candidate: ", candidate)
        if candidate < minimum and search_engine != current:
            minimum = candidate
            answer_engine = search_engine

    return answer_engine

N = int(input())
N_begin = 1
while (N_begin<=N):
    N_begin += 1
    S = input()
    S_cp = int(S)
    S_list = []

    while (S_cp>0):
        S_list.append(input())
        S_cp -= 1
        
    Q = input()
    Q_cp = int(Q)
    Q_list = []
    
    while (Q_cp>0):
        Q_list.append(input())
        Q_cp -= 1

    min_switches = 0

    current_engine = find_min_search_engine(S_list, Q_list, 0, "")
    # O(N^2), probably can improve more later
    while (Q_cp < int(Q)-1): # Currently Q_cp is zero
        print("current best engine: ", current_engine)
        # Finding a search engine that has the min number of queries in Q_list
        if Q_list[Q_cp] == current_engine:
            current_engine = find_min_search_engine(S_list, Q_list, Q_cp, current_engine)
            min_switches += 1
        Q_cp += 1

    print("Case #", N_begin, ": ", min_switches)

